### PR TITLE
Fix fp8 moe reverse substring

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Fix DSv4 / FP8 inits. 

By construction FP8 gate_up_proj scales need 2 tensor to save a dummy model with revert conversion!

### CODE REVIEW

File: src/transformers/models/deepseek_v4/configuration_deepseek_v4.py [hunk 1]
```diff
@@ -304,9 +310,14 @@ def __post_init__(self, **kwargs):
                 **yarn,
                 "rope_theta": self.compress_rope_theta,
                 "partial_rotary_factor": self.partial_rotary_factor,
-                "attention_factor": 1.0,
             }
             compress.setdefault("rope_type", "default")
+            # `attention_factor=1.0` only does something under yarn (it suppresses
+            # `_compute_yarn_parameters`'s `0.1·log(factor)+1` mscale). Adding it on
+            # rope_type=default produces an "Unrecognized key" warning for no benefit —
+            # happens on re-constructions where yarn params have already been consumed.
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v4/configuration_deepseek_v4.py]:
```suggestion
```

File: src/transformers/models/deepseek_v4/modeling_deepseek_v4.py [hunk 2]
```diff
@@ -412,9 +412,10 @@ def forward(
         if chunk_kv.shape[1] > 0:  # there were at least self.compress_rate tokens
             n_windows = chunk_kv.shape[1] // self.compress_rate
             chunk_kv = chunk_kv.view(batch, n_windows, self.compress_rate, -1)
-            chunk_gate = chunk_gate.view(batch, n_windows, self.compress_rate, -1) + self.position_bias.to(
-                chunk_gate.dtype
-            )
+            # Reference keeps the gate in fp32 from here through the softmax (Compressor.forward
+            # in DeepSeek-V4-Flash/inference/model.py upcasts x to fp32 at entry). bf16 + fp32
+            # broadcasts to fp32; we deliberately don't down-cast so new_gate stays fp32 too.
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v4/modeling_deepseek_v4.py]:
```suggestion
```

File: src/transformers/models/deepseek_v4/modeling_deepseek_v4.py [hunk 3]
```diff
@@ -515,7 +516,8 @@ def forward(
             n_windows = chunk_kv.shape[1] // self.compress_rate
             ratio = self.compress_rate
             chunk_kv = chunk_kv.view(batch, n_windows, ratio, -1)
-            chunk_gate = chunk_gate.view(batch, n_windows, ratio, -1) + self.position_bias.to(chunk_gate.dtype)
+            # bf16 + fp32 broadcasts to fp32; keep it fp32 through the softmax (matches reference).
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v4/modeling_deepseek_v4.py]:
```suggestion
```

File: src/transformers/models/deepseek_v4/modeling_deepseek_v4.py [hunk 4]
```diff
@@ -633,7 +635,8 @@ def forward(
...
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v4/modeling_deepseek_v4.py]:
```suggestion
```

File: src/transformers/models/deepseek_v4/modular_deepseek_v4.py [hunk 5]
```diff
@@ -348,9 +348,10 @@ def forward(
         if chunk_kv.shape[1] > 0:  # there were at least self.compress_rate tokens
             n_windows = chunk_kv.shape[1] // self.compress_rate
             chunk_kv = chunk_kv.view(batch, n_windows, self.compress_rate, -1)
-            chunk_gate = chunk_gate.view(batch, n_windows, self.compress_rate, -1) + self.position_bias.to(
-                chunk_gate.dtype
-            )
+            # Reference keeps the gate in fp32 from here through the softmax (Compressor.forward
+            # in DeepSeek-V4-Flash/inference/model.py upcasts x to fp32 at entry). bf16 + fp32
+            # broadcasts to fp32; we deliberately don't down-cast so new_gate stays fp32 too.
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v4/modular_deepseek_v4.py]:
```suggestion
```

File: src/transformers/models/deepseek_v4/modular_deepseek_v4.py [hunk 6]
```diff
@@ -451,7 +452,8 @@ def forward(
             n_windows = chunk_kv.shape[1] // self.compress_rate
             ratio = self.compress_rate
             chunk_kv = chunk_kv.view(batch, n_windows, ratio, -1)
-            chunk_gate = chunk_gate.view(batch, n_windows, ratio, -1) + self.position_bias.to(chunk_gate.dtype)
+            # bf16 + fp32 broadcasts to fp32; keep it fp32 through the softmax (matches reference).
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v4/modular_deepseek_v4.py]:
```suggestion
```

File: src/transformers/models/deepseek_v4/modular_deepseek_v4.py [hunk 7]
```diff
@@ -569,7 +571,8 @@ def forward(
...
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v4/modular_deepseek_v4.py]:
```suggestion
```

File: src/transformers/core_model_loading.py [hunk 8]
```diff
@@ -927,6 +956,34 @@ def __init__(
...
+    def reverse_transform(self) -> WeightTransform:
+        # Append `(?=\.|$)` to each reversed source so the match has to end on a
+        # dot-token boundary. Stops `mlp.experts.gate_up_proj` from substring-matching
+        # the FP8 scale sibling `mlp.experts.gate_up_proj_scale_inv` on save.
+
+        # `quantization_operation` is the on-the-fly quantize op attached by
+        # `convert_and_load_state_dict_in_model` when the checkpoint wasn't pre-quantized.
+        # We only allow reversing if the op has been opted in via `supports_round_trip`
+        # (audited reverse_op pair, e.g. Fp8Quantize ↔ Fp8Dequantize). Unaudited ops
+        # bail with a clear error rather than silently writing a half-converted checkpoint.
```
Comment (ArthurZucker) [inline src/transformers/core_model_loading.py]:
```suggestion
```

File: src/transformers/core_model_loading.py [hunk 9]
```diff
@@ -927,6 +956,34 @@ def __init__(
...
+        if self.quantization_operation is not None and not self.quantization_operation.supports_round_trip:
```
Comment (ArthurZucker) [inline src/transformers/core_model_loading.py]:
```suggestion
        if self.quantization_operation is not None and getattr(self.quantization_operation, "reverse_op", None) is not None:
```

File: src/transformers/core_model_loading.py [hunk 10]
```diff
@@ -83,6 +83,12 @@ def build_glob_alternation(
 class ConversionOps:
     """Base class for weight conversion operations."""
 
+    # Opt-in: set True on subclasses whose `reverse_op` has been audited to round-trip
+    # cleanly (forward(reverse(x)) == x within numerical tolerance). Default False keeps
+    # the save path conservative — unaudited ops bail rather than silently corrupt a
+    # checkpoint. The bail lives in `WeightConverter.reverse_transform`.
+    supports_round_trip: bool = False
```
Comment (ArthurZucker) [inline src/transformers/core_model_loading.py]:
```suggestion
```

File: src/transformers/core_model_loading.py [hunk 11]
```diff
@@ -83,6 +83,15 @@ def build_glob_alternation(
...
     """Base class for weight conversion operations."""
 
+    # Quantization-only opt-in: set True on a quantize/dequantize op once its `reverse_op`
+    # has been audited to round-trip a save cycle correctly. Today only `Fp8Quantize` /
+    # `Fp8Dequantize` are flagged — every other quantization backend (bnb, torchao, eetq,
+    # mxfp4, quanto, ...) stays False until verified. `WeightConverter.reverse_transform`
+    # only consults this on the converter's `quantization_operation`; the regular shape
+    # ops (Chunk, MergeModulelist, ...) reverse themselves through their own `reverse_op`
+    # property and don't need a flag.
+    supports_round_trip: bool = False
```
Comment (ArthurZucker) [inline src/transformers/core_model_loading.py]:
```suggestion
```

File: src/transformers/core_model_loading.py [hunk 12]
```diff
@@ -927,6 +939,34 @@ def __init__(
...
+        # Append `(?=\.|$)` to each reversed source so the match has to end on a
+        # dot-token boundary. Stops `mlp.experts.gate_up_proj` from substring-matching
+        # the FP8 scale sibling `mlp.experts.gate_up_proj_scale_inv` on save.
+
+        # `quantization_operation` is the on-the-fly quantize op attached by
+        # `convert_and_load_state_dict_in_model` when the checkpoint wasn't pre-quantized.
+        # Only opt-in (`supports_round_trip=True`) quant ops are safe to reverse — today
+        # that's just `Fp8Quantize` / `Fp8Dequantize`. Unaudited quant backends bail
+        # rather than silently write a half-converted checkpoint.
+        if self.quantization_operation is not None and not self.quantization_operation.supports_round_trip:
```
Comment (ArthurZucker) [inline src/transformers/core_model_loading.py]:
```suggestion
        if self.quantization_operation is not None and not getattr(self.quantization_operation, "supports_round_trip", False):
```

File: src/transformers/core_model_loading.py [hunk 13]
```diff
@@ -927,6 +939,34 @@ def __init__(
...
+    def reverse_transform(self) -> WeightTransform:
+        # Append `(?=\.|$)` to each reversed source so the match has to end on a
+        # dot-token boundary. Stops `mlp.experts.gate_up_proj` from substring-matching
+        # the FP8 scale sibling `mlp.experts.gate_up_proj_scale_inv` on save.
+
+        # `quantization_operation` is the on-the-fly quantize op attached by
+        # `convert_and_load_state_dict_in_model` when the checkpoint wasn't pre-quantized.
+        # Only opt-in (`supports_round_trip=True`) quant ops are safe to reverse — today
+        # that's just `Fp8Quantize` / `Fp8Dequantize`. Unaudited quant backends bail
+        # rather than silently write a half-converted checkpoint.
```
Comment (ArthurZucker) [inline src/transformers/core_model_loading.py]:
```suggestion
```

Comment (ArthurZucker) [inline src/transformers/core_model_loading.py]:
```suggestion
```

**Comment by qgallouedec** [comment]
thanks! regression from #45930, I tested this branch with https://github.com/huggingface/transformers/pull/45930#issuecomment-4444187780, it works:

```
$ /fsx/qgallouedec/miniconda3/envs/trl/bin/python /fsx/qgallouedec/transformers/reg.py
Writing model shards: 100%|██████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 18.66it/s]
[transformers] You have loaded an FP8 model on CPU and have a CUDA or XPU device available, make sure to set your model on a GPU or XPU device in order to run your model. To remove this warning, pass device_map = 'cuda' or 'xpu'. 
Loading weights: 100%|████████████████████████████████████████████████████████████████████████████████| 84/84 [00:00<00:00, 751.16it/s]
model.embed_tokens.weight torch.Size([128, 8]) torch.bfloat16
model.layers.0.self_attn.sinks torch.Size([4]) torch.float32
model.layers.0.self_attn.q_a_proj.weight torch.Size([1024, 8]) torch.float8_e4m3fn
model.layers.0.self_attn.q_a_proj.weight_scale_inv torch.Size([8, 1]) torch.float32
model.layers.0.self_attn.q_a_norm.weight torch.Size([1024]) torch.float32
model.layers.0.self_attn.q_b_proj.weight torch.Size([2048, 1024]) torch.float8_e4m3fn
model.layers.0.self_attn.q_b_proj.weight_scale_inv torch.Size([16, 8]) torch.float32
model.layers.0.self_attn.kv_proj.weight torch.Size([512, 8]) torch.float8_e4m3fn
model.layers.0.self_attn.kv_proj.weight_scale_inv torch.Size([4, 1]) torch.float32
model.layers.0.self_attn.kv_norm.weight torch.Size([512]) torch.float32
model.layers.0.self_attn.o_a_proj.weight torch.Size([8192, 256]) torch.float8_e4m3fn
model.layers.0.self_attn.o_a_proj.weight_scale_inv torch.Size([64, 2]) torch.float32
model.layers.0.self_attn.o_b_proj.weight torch.Size([8, 8192]) torch.float8_e4m3fn
model.layers.0.self_attn.o_b_proj.weight_scale_inv torch.Size([1, 64]) torch.float32
model.layers.0.self_attn.compressor.position_bias torch.Size([128, 512]) torch.float32
model.layers.0.self_attn.compressor.kv_proj.weight torch.Size([512, 8]) torch.float8_e4m3fn
model.layers.0.self_attn.compressor.kv_proj.weight_scale_inv torch.Size([4, 1]) torch.float32
model.layers.0.self_attn.compressor.gate_proj.weight torch.Size([512, 8]) torch.float8_e4m3fn
model.layers.0.self_attn.compressor.gate_proj.weight_scale_inv torch.Size([4, 1]) torch.float32
model.layers.0.self_attn.compressor.kv_norm.weight torch.Size([512]) torch.float32
model.layers.0.mlp.gate.weight torch.Size([256, 8]) torch.bfloat16
model.layers.0.mlp.experts.gate_up_proj torch.Size([256, 64, 8]) torch.bfloat16
model.layers.0.mlp.experts.gate_up_proj_scale_inv torch.Size([256, 1, 1]) torch.float32
model.layers.0.mlp.experts.down_proj torch.Size([256, 8, 32]) torch.bfloat16
model.layers.0.mlp.experts.down_proj_scale_inv torch.Size([256, 1, 1]) torch.float32
model.layers.0.mlp.shared_experts.gate_proj.weight torch.Size([32, 8]) torch.float8_e4m3fn
model.layers.0.mlp.shared_experts.gate_proj.weight_scale_inv torch.Size([1, 1]) torch.float32
model.layers.0.mlp.shared_experts.up_proj.weight torch.Size([32, 8]) torch.float8_e4m3fn
model.layers.0.mlp.shared_experts.up_proj.weight_scale_inv torch.Size([1, 1]) torch.float32
model.layers.0.mlp.shared_experts.down_proj.weight torch.Size([8, 32]) torch.float8_e4m3fn
model.layers.0.mlp.shared_experts.down_proj.weight_scale_inv torch.Size([1, 1]) torch.float32
model.layers.0.input_layernorm.weight torch.Size([8]) torch.float32
model.layers.0.post_attention_layernorm.weight torch.Size([8]) torch.float32
model.layers.0.attn_hc.fn torch.Size([24, 32]) torch.float32
model.layers.0.attn_hc.base torch.Size([24]) torch.float32
model.layers.0.attn_hc.scale torch.Size([3]) torch.float32
model.layers.0.ffn_hc.fn torch.Size([24, 32]) torch.float32
model.layers.0.ffn_hc.base torch.Size([24]) torch.float32
model.layers.0.ffn_hc.scale torch.Size([3]) torch.float32
...
```

File: src/transformers/integrations/finegrained_fp8.py [hunk 14]
```diff
@@ -989,7 +990,10 @@ def _dequantize_one(self, quantized: torch.Tensor, scales: torch.Tensor) -> torc
         # Derive block size from the scale grid rather than the global config: MoE experts
         # ship MXFP4 with a ``[1, 32]`` block, dense linears ship FP8 with ``[128, 128]``,
         # and the same dequant has to handle both within one checkpoint.
-        scale_rows, scale_cols = scales.shape[-2:]
+        try:
+            scale_rows, scale_cols = scales.shape[-2:]
+        except Exception as e:
```
Comment (vasqu) [inline src/transformers/integrations/finegrained_fp8.py]: In which situation do we get the error? Maybe a slight comment just to clarify its really expected

File: src/transformers/models/deepseek_v4/modular_deepseek_v4.py [hunk 15]
```diff
@@ -569,7 +567,7 @@ def forward(
             n_windows = chunk_kv.shape[1] // self.compress_rate
             ratio = self.compress_rate
             chunk_kv = chunk_kv.view(batch, n_windows, ratio, -1)
-            chunk_gate = chunk_gate.view(batch, n_windows, ratio, -1) + self.position_bias.to(chunk_gate.dtype)
```
Comment (vasqu) [inline src/transformers/models/deepseek_v4/modular_deepseek_v4.py:572]: Any reason we dont need the type casts anymore? Seems like it affects generations, no?

File: src/transformers/conversion_mapping.py [hunk 16]
```diff
@@ -416,12 +416,12 @@ def _build_checkpoint_conversion_mapping():
...
                     "mlp.experts.*.w3.weight",
                 ],
-                target_patterns="mlp.experts.gate_up_proj$",
+                target_patterns="mlp.experts.gate_up_proj",
                 operations=[MergeModulelist(dim=0), Concatenate(dim=1)],
             ),
             WeightConverter(
                 source_patterns="mlp.experts.*.w2.weight",
-                target_patterns="mlp.experts.down_proj$",
+                target_patterns="mlp.experts.down_proj",
```
Comment (vasqu) [inline src/transformers/conversion_mapping.py:424]: Nice!

File: src/transformers/core_model_loading.py [hunk 17]
```diff
@@ -123,16 +123,13 @@ def convert(
...
         tensor = tensors[0] if isinstance(tensors, list) else tensors
-        targets = self.get_target_patterns(input_dict, target_patterns)
+        targets = target_patterns
         sizes = len(targets)
         chunks = torch.chunk(tensor, sizes, dim=self.dim)
+        if len(input_dict) > 1 or len(target_patterns) == 1 or len(chunks) != len(target_patterns):
+            raise ValueError(f"Failed to convert {kwargs.get('full_layer_name')}")
         return dict(zip(targets, chunks))
 
-    def get_target_patterns(self, input_dict: dict, target_patterns: list[str]) -> list[str]:
```
Comment (vasqu) [inline src/transformers/core_model_loading.py:131]: Imo, would keep this but add the additional condition there. I think we have several of these now so its just style

`sizes = len(target_patterns)` then

**Review by vasqu — APPROVED** [review]
Mostly nits, are the dtype casts expected to be removed?

File: src/transformers/models/deepseek_v4/modular_deepseek_v4.py (hunk 15 above)
Comment (ArthurZucker) [reply src/transformers/models/deepseek_v4/modular_deepseek_v4.py:572]: It fixes the generation as we accumualate + softmax on fp32˜

File: src/transformers/core_model_loading.py (hunk 17 above)
Comment (ArthurZucker) [reply src/transformers/core_model_loading.py:131]: okay!

Comment (ArthurZucker) [reply src/transformers/core_model_loading.py:131]: ah but I need the chunks, the kwargs etc

Comment (vasqu) [reply src/transformers/core_model_loading.py:131]: Hmm, can just pass them, no?
